### PR TITLE
Move redis adapter to separate npm library

### PR DESCRIPTION
**Pull request by mannyv123**
### Description
This change removes the redis 4 adapter from `@epic-web/cachified` and links to a separate npm library I set up:
[cachified-redis-adapter](https://github.com/mannyv123/cachified-redis-adapter)

### Why
This helps avoid issues when updating the adapter and causing breaking change releases for the entire `@epic-web/cachified` package.

### Details
- `cachified-redis-adapter` basically takes the code that was in `@epic-web/cachified`  for the redis adapter and moves it into a separate library
- this PR then updates `@epic-web/cachified` to remove the redis adapter code and update the `readme.md` file to link to `cachified-redis-adapter`

**Note**: I wasn't too sure how to update the `extractExamples.js` file. I can see that since the `readme.md` file is update it has removed the redis adapter example from generating. Any suggestions here would be appreciated!

@tearingItUp786

### CODE REVIEW

**Review by Xiphe — COMMENTED** [review]
Code-wise: LGTM

But we need to decide if we want to release this as a breaking change or not.

- A) When we want to actually remove the adapter:   
  Add `BREAKING CHANGE` to one of the commit messages (or to the squash merge message) in order to issue a new major version. See [semantic-release#commit-message-format](https://github.com/semantic-release/semantic-release?tab=readme-ov-file#commit-message-format) for details.
- B) When we don't want to create a new major version just yet:  
  Keep the adapter in and `@deprecate`.  The docs can be updated to point towards the new package.

Leaving this decision to you @kentcdodds

**Comment by mannyv123** [comment]
Thanks everyone for the feedback! I've made the following changes:
- removed redis3 adapter code and related tests
- removed `redis-mock`, `@types/redis-mock`, `redis4` packages

I also added `BREAKING CHANGE` to the commit message, hopefully formatted it correctly🤞

Thanks again!

**Review by kentcdodds — CHANGES_REQUESTED** [review]
Actually, the more I think about it, the more I think I would prefer to delete all adapters. And I think the README could be cleaned up as well. Do you have time and interest in doing that please @mannyv123?

**Comment by tearingItUp786** [comment]
Hey hey! I'll probably be helping @mannyv123 with this 😄 

So the plan for the work going forward: 
- remove `adapters.ts` 
- Update the `README`
- Update the `extractExamples.js` or do we want to remove this?

**Comment by kentcdodds** [comment]
I think we can remove that. Thanks @tearingItUp786!

**Comment by Xiphe** [comment]
Voting to keep `extractExamples.js` but as the creator of it I'm certainly biased :)

I've created that one because I tend to forget keeping (readme) examples up to date and wanted to have a safeguard. 
Let me know when you want my help with it.

**Comment by kentcdodds** [comment]
I think it's a cool idea, but with removing all the adapters and moving the docs for them into their respective packages combined with the level of complexity there, I don't think it makes sense to keep it around.

**Comment by Xiphe** [comment]
The adapters were actually the examples that have not been tested :) 
But I'm completely fine with removing it 👍

**Comment by mannyv123** [comment]
Hi all!

I've made the changes as discussed:
- removed adapters and tests
- removed `extractExamples.js` and related tests
- updated the `README`

@kentcdodds you mentioned during the last office hours that it would be good to update the usage example in the `README` with the lru adapter code and that we might be able to reduce the amount of code as well. I updated the example with the lru adapter but wasn't too sure how to make it simpler. Would you be able to provide some feedback/guidance on how to update this? Thank you!

File: README.md [hunk 1]
```diff
@@ -52,10 +52,40 @@ npm install @epic-web/cachified
...
+    get(key) {
+      return lruCache.get(key);
+    },
+    delete(key) {
+      return lruCache.delete(key);
+    },
+  };
+}
+
+const lru = lruCacheAdapter(lruInstance);
```
Comment (kentcdodds) [inline README.md:88]: Remove this stuff and just set `lru` to the cache instance itself. So something like:

```
const lru: Cache = {
  set() {},
  get() {},
  delete() {},
}
```

Of course you'll want to fill that in with working code and make sure the types are happy too. I think you'll want to use the Cache generic maybe.

**Review by kentcdodds — COMMENTED** [review]
This is looking good to me. Definitely simplifies things for us development-wise and reduces the chance that we'll need breaking changes in the future.

Just one thing on the example in the README and then I'll be ready to merge this.

**Review by Xiphe — APPROVED** [review]
Agree with Kents feedback. Rest: LGTM!

**Comment by mannyv123** [comment]
Thank you for the feedback!

I've updated the usage example now. Please let me know if this is in line with what you were thinking or if any changes are needed. 

Thanks again!

File: README.md [hunk 2]
```diff
@@ -52,40 +52,23 @@ npm install @epic-web/cachified
...
+  set(key, value) {
+    return lruInstance.set(key, value);
+  },
+  get(key) {
+    return lruInstance.get(key);
+  },
+  delete(key) {
+    return lruInstance.delete(key);
+  },
+};
```
Comment (Xiphe) [inline README.md:71]: This implementation has no benefit over

```ts
const cache = new LRUCache<string, CacheEntry>({ max: 1000 })

cachified({ cache, /* ... */ });
```

One of the main reasons for an adapter is to set the TTL on the cached item in a way that the cache will clean it up itself without involving `cachified`.

Proposing to either not create an on the fly adapter (example above) or do the following:

```ts
/* lru cache is not part of this package but a simple non-persistent cache */
const lruInstance = new LRUCache<string, CacheEntry>({ max: 1000 });

const cache: Cache = {
  set(key, value) {
    const ttl = totalTtl(value?.metadata);
    return lruCache.set(key, value, {
      ttl: ttl === Infinity ? undefined : ttl,
      start: value?.metadata?.createdTime,
    });
  },
  get(key) {
    return lruInstance.get(key);
  },
  delete(key) {
    return lruInstance.delete(key);
  },
};
```

Comment (Xiphe) [reply README.md:71]: Would vote for the latter option. In real apps this is quite critical in order to not bloat your cache db.

Comment (mannyv123) [reply README.md:74]: Ahh that makes sense! Thank you for the help!

I've updated it now to the above example.

**Comment by mannyv123** [comment]
Hi all,

I resolved the merge conflicts that were in the package.json file. 

Also, I made an update to the `tsconfig.json` file as when `tsc` was run, it was no longer creating the `dist/src` folder structure. It seems this was caused when I removed the `readme.test.ts` file. Please let me know if my change here was incorrect and I can update.

Thanks!

**Review by Xiphe — APPROVED** [review]
LGTM :whale2:

Good catch with the `src` folder. The other option would have been to remove the `src/` in the `"type"` paths in package.json but I'm fine with either 👍

**Comment by caiocmpaes** [comment]
Hi, everyone! 👋 

I just updated the package to `5.1.1` and then seems like we need a bump on the peer dependency of [cachified-redis-adapter](https://www.npmjs.com/package/cachified-redis-adapter).

Captura de Tela 2024-02-09 às 21 48 04

**Comment by tearingItUp786** [comment]
> Hi, everyone! 👋
> 
> I just updated the package to `5.1.1` and then seems like we need a bump on the peer dependency of [cachified-redis-adapter](https://www.npmjs.com/package/cachified-redis-adapter).
> 
> Captura de Tela 2024-02-09 às 21 48 04

https://github.com/mannyv123/cachified-redis-adapter/pull/7

**Comment by mannyv123** [comment]
@caiocmpaes @tearingItUp786 PR is now merged! :)
Thank you!

**Pull request merged**